Fix Snackbar overlay not covering any open modals, plus keyboard a11y fixes

jonboiser opened the pull request:
### Summary

1. Addresses #6301 by simply making the Snackbar overlay div one z-index level higher than the KModal overlay div.
1. Adds a keyboard-focus outline to the snackbar action button and fixes an issue where focus would be trapped on the snackbar div, making the action button inaccessible when the snackbar backdrop was applied.


### Reviewer guidance

1. Test out the scenario in #6301 (e.g. open a modal and turn off the server). Make sure things still work when using keyboard only as well as mouse.
1. To test out the keyboard accessibility in other context, try accessing the snackbar while editing lessons (since the snackbar will have an "undo" action) when you remove items.

### References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

…

----

### Contributor Checklist


PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [x] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

jonboiser committed:
Fix KModal overlay being masked by Snackbar overlay and the focus infinite loop

jonboiser committed:
Give snackbar action button keyboard focus outline

jonboiser committed:
Add better guard against infinite recursion when focus is given to snackbar

bjester commented:
When I click the `Try Now` button on the offline snackbar, I still get tons of these errors: `RangeError: Maximum call stack size exceeded.`

jonboiser commented:
@bjester Which browser are you using, and which modal did you open before disconnecting the server?

bjester commented:
@jonboiser Chrome 80.0.3987.122 and I was using the confirmation modal that appears when deleting downloaded content nodes.

jonboiser commented:
Hm, just tried the test scenario in the same area of the app and I don't get the errors. Maybe it's a Linux thing. @nucleogenesis Could you see if something similar happens for you?

jonboiser commented:
If anything, the culprit might be cross-browser/OS differences in how the DOM or Event API is implemented, which is crucial to the `noopOnFocus` variable that prevents the focus/refocusing infinite recursion

bjester commented:
@jonboiser Yeah it must be related to some browser differences, because I don't experience that error on Firefox.

jonboiser commented:
Could you put a debugger at this line https://github.com/jonboiser/kolibri/blob/1bfd842b6fcd326c184ad3a91855a22e8e10e673/packages/kolibri-components/src/KModal.vue#L278

and see what element the `event.target` is when you get the error?

bjester commented:
`event.target` is `window`

jonboiser commented:
That's weird. Maybe the check `event.target === window` doesn't actually work?

bjester commented:
That does work, `e.target === window`. I should probably say that I'm using the pex from your PR build here too, in case that is somehow broken.

jonboiser commented:
Okay, I can finally reproduce this on the PEX 🤷‍♂

bjester commented on packages/kolibri-components/src/KModal.vue:
```diff
@@ -312,7 +307,8 @@
     left: 0;
-    z-index: 24;
+    // z-index needs to be one less than .snackbar-backdrop
+    z-index: 15;
```
I know @indirectlylit has informed me multiple times that elevations should align with [MD elevations](https://material.io/design/environment/elevation.html#default-elevations). It says snackbars are 6 and dialogs are 24. That doesn't really help but I don't know how to resolve the discrepancy.

jonboiser replied:
Hm, I think elevations aren't really z-indices, but actually correspond to actual elevations in display points and usually affected with box-shadow rules (plus obviously having the right order of magnitude of z-index). I guess it's just a coincidence that the z-indices match up with the elevations.

indirectlylit replied:
It's not a coincidence, no - but it's also not a Material spec

We chose by convention to align Kolibri z-indexes with Material elevations to avoid the situation where people choose arbitrary numbers just to get something above or below something else.

I think it's fine to have exceptions to that convention as long as we label them as-such, but in general let's try to stick to matching MD elevations.

jonboiser commented:
Okay, IDK why it works differently with the pex, but I think what's happening is that the focus event is propagating down to the "Try again" button, which is not caught by my `noopOnFocus` thing. So I'm going to add yet another condition that will tell KModal to stop trying to focus itself if focus is given to the snackbar or any of its children via

```
const noopOnFocus = ... || 
event.target.id === "coresnackbar" ||		
document.getElementById("coresnackbar").contains(event.target)
```

jonboiser committed:
Use a div to trap focus in snackbar instead of using a listener on window

jonboiser commented:
@bjester @nucleogenesis Besides fixing the guard in KModal I made a change to the snackbar to use a slighty lighter approach to trapping the focus, kind of like what I did for the CoreMenu since it reacting to focus events over all of window was giving me problems. Could you give it another try with the PEX?

bjester commented:
Editing content, I'm able to access the snackbar and its action button using the keyboard. The issue I experienced previously is also fixed. So the entire snackbar should still receive focus, but should allow tabbing to the action button if there is one?

bjester commented:
@jonboiser In Chrome, the focus trapping worked fine when editing content, but in Firefox it isn't working correctly. The snackbar doesn't get focus

jonboiser commented:
> So the entire snackbar should still receive focus, but should allow tabbing to the action button if there is one?

That's how it was initially implemented (with tabindex=0 on the snackbar div), but on the keyboard modality, this ends up showing the focus ring on the whole thing. I think ideally, I would only want to tab into the action button if the snackbar has one (and not tab the snackbar around it), otherwise the whole snackbar has no tabindex.

jonboiser committed:
Add another focus trap underneath the snackbar

jonboiser commented:
Okay, I made another change that will hopefully help with FF. 

1. I added another focus trap underneath the action button, which will be active whether there is a backdrop or not. I personally found that the ability to tab out of the app and into the browser chrome (into the dev tools or who knows where) was sort of confusing anyway. I tested in safari, chrome, ff with the disconnect modal and don't think i can remove the focus from the snackbar with mouse or keyboard.
1. I removed the focus outlining from the snackbar div. So you can tab into it with the keyboard, but it won't receive an outline, tab again to focus the button.

bjester commented:
I still can't get focus on the snackbar in FF :/

jonboiser commented:
I think the issue on FF might be related to the DND widget on the lesson's page. I experience the same thing where no matter where I click on it/drag it/etc, the focus is immediately thrown to the top of the page I have to race to tab into the snackbar. (But weirdly this is only on the PEX installer, seems fine running the devserver).

I think this is the only part of the app where this is an issue (and should probably be changed because it's not good for a11y). But overall it seems this PR fixes the original disconnect snackbar + modal issue.
